File: solution/c#/Day19/Day19/Blog.cs
using System.Collections.Immutable;
using LanguageExt;

namespace Day19
{
    public class Article
    {
        private readonly string _name;
        private readonly string _content;
        public ImmutableArray<Comment> Comments { get; }

        private Article(string name, string content, IEnumerable<Comment> comments)
        {
            _name = name;
            _content = content;
            Comments = comments.ToImmutableArray();
        }

        public Article(string name, string content)
            : this(name, content, Array.Empty<Comment>())
        {
        }

        public Either<Error, Article> AddComment(string text, string author)
        {
            var comment = new Comment(text, author, Now());

            return Comments.Contains(comment)
                ? new Error("This comment already exists in this article")
                : new Article(_name, _content, Comments.Append(comment));
        }

        private static DateOnly Now() => DateOnly.FromDateTime(DateTime.Now);
    }

    public record Error(string Message);

    public record Comment(string Text, string Author, DateOnly CreationDate);
}

File: solution/ts/day13/src/blog.ts
export class Article {
    private readonly name: string;
    private readonly content: string;
    private readonly comments: Comment[];

    constructor(name: string, content: string) {
        this.name = name;
        this.content = content;
        this.comments = [];
    }

    private addCommentWithCreationDate = (text: string, author: string, creationDate: Date): void => {
        const comment = new Comment(text, author, creationDate);

        if (this.comments.some(c => c.isSame(comment))) {
            throw new Error("Comment already exists");
        } else {
            this.comments.push(comment);
        }
    };

    public addComment = (text: string, author: string): void => this.addCommentWithCreationDate(text, author, new Date());
    public getComments = (): Comment[] => this.comments;
}

export class Comment {
    constructor(
        public readonly text: string,
        public readonly author: string,
        public readonly creationDate: Date
    ) {}

    isSame = (otherComment: Comment) => this.text === otherComment.text
        && this.author === otherComment.author
        && this.creationDate.getDate() === otherComment.creationDate.getDate();
}
